### PR TITLE
Search Feature and Documentation related bug fix

### LINKED ISSUE #162: [PE-D][Tester E] Inconsistency in command format in the UG and the application
I propose that only one is followed.

-------------
Labels: `severity.Low` `type.DocumentationBug`
original: yixiann/ped#4

### LINKED ISSUE #170: [PE-D][Tester E] Name disallows a valid legal name containing s/o
This is from the CS2103T website

-------------
Labels: `type.FunctionalityBug` `severity.Medium`
original: yixiann/ped#6

### LINKED ISSUE #173: [PE-D][Tester E] Guard for duplicates is not strong enough to guard against close matches
-------------
Labels: `type.FunctionalityBug` `severity.Medium`
original: yixiann/ped#7

### LINKED ISSUE #175: [PE-D][Tester E] What if individuals have the same name? Does this mean we cannot add them both into SoConnect?
No details provided by bug reporter.

-------------
Labels: `severity.VeryLow` `type.FeatureFlaw`
original: yixiann/ped#8

### LINKED ISSUE #182: [PE-D][Tester B] Add command does not allow addition of a new contact with the same name but with all other fields different
It is quite common for people to have the same name. So to check whether the person being added is the same person, you can check for the other fields also, such as phone number, email addresses. 

E.g., if I have two friends named Charles, then this software does not allow me to add the two Charles in the address book.

-------------
Labels: `type.FeatureFlaw` `severity.Medium`
original: PeiYee88/ped#7

### LINKED ISSUE #185: [PE-D][Tester B] Numerical values accepted for name field
Perhaps can be stricter, e.g., only accepts alphabets. 

-------------
Labels: `type.FunctionalityBug` `severity.Medium`
original: PeiYee88/ped#8

### LINKED ISSUE #188: [PE-D][Tester B] Consideration in accepting comma in name field
Could not add my friend's name with the add command because he has a comma in his name (Ong Jing Kai, William). So might need to take that into consideration.

Currently it only accepts alphanumeric value

-------------
Labels: `type.FeatureFlaw` `severity.Low`
original: PeiYee88/ped#9

### LINKED ISSUE #198: [PE-D][Tester B] Edit command does not allow edit to be done on contact with the same name
In my opinion, I think edit should be allowed unless all the other fields are the same. Quite common to have the same name

-------------
Labels: `type.FeatureFlaw` `severity.Medium`
original: PeiYee88/ped#11

### LINKED ISSUE #200: [PE-D][Tester A] Search bug
input: search n/al

expected: either all name with al comes out or none.

actual: irfan ibrahim came out

Could be a nice feature if you can search part of the name only.

-------------
Labels: `severity.Medium` `type.FunctionalityBug`
original: LolfoollorS/ped#8

### LINKED ISSUE #210: [PE-D][Tester A] Cannot add a different person with same name
If there is two person with same name but have different phone, address and email, this program will not accept it.

Reason being that there are alot of people in the world with the same name even in facebook and google has many people with the same name.

-------------
Labels: `type.FeatureFlaw` `severity.Low`
original: LolfoollorS/ped#10

### PR DESCRIPTION
**Pull request by florentianayuwono**

### CODE REVIEW

**Comment by florentianayuwono** [comment]
Some search issues cannot be linked cuz it only allows max 10 linked issues. Probably i will linked the rest search issues at the next pr about UI fix

File: src/main/java/soconnect/logic/parser/SearchCommandParser.java [hunk 1]
```diff
@@ -84,13 +84,14 @@ private SearchCommand parseSearchCondition(ArgumentMultimap argMultimap, String
...
+                new ContactMightBeRelevantPredicate(argMultimap, false);
+        Predicate<Person> contactMightBeLessRelevantPredicate = new ContactMightBeRelevantPredicate(argMultimap, true);
         if (isJointCondition) {
             return new SearchCommand(new ContactContainsAllKeywordsPredicate(argMultimap),
-                    contactMightBeRelevantPredicate);
+                    contactMightBeRelevantPredicate, contactMightBeLessRelevantPredicate);
         } else {
             return new SearchCommand(new ContactContainsAnyKeywordsPredicate(argMultimap),
-                    contactMightBeRelevantPredicate);
+                    contactMightBeRelevantPredicate, contactMightBeLessRelevantPredicate);
```
Comment (Eugene-Ong-W-X) [inline src/main/java/soconnect/logic/parser/SearchCommandParser.java:98]: For better SLAP, you can abstract out `new ContactContainsAllKeywordsPredicate(argMultimap)` and `new ContactContainsAnyKeywordsPredicate(argMultimap)` from your calling of the SearchCommand constructor.

**Review by Eugene-Ong-W-X — COMMENTED** [review]
Great work on the changes! I just have 2 questions
1. When testing `search n/al` on the sample data, out of the 4 results, I noticed that `David Li` and `Charlotte` were returned. If I understood the code correctly, this is because for relevant search results, they are returned based on matching character by character? So the 2 mentioned results appear because they contain `a` and `l` separately. If so, I feel there's a need to update the UG to explain more clearly how relevant search results work so users won't be confused on why certain search results appear. (i.e. they won't be mistaken that the search mechanism is not working correctly)

2. Does this PR fix the bug for single letter tag searches like `search t/t` or `search t/x`? Because I tested it and `search t/x` still shows all 6 contacts. If the bug isn't meant to be fixed in this PR, you can ignore this, just checking 👍

File: src/main/java/soconnect/logic/parser/SearchCommandParser.java (hunk 1 above)
Comment (florentianayuwono) [reply src/main/java/soconnect/logic/parser/SearchCommandParser.java:98]: Good suggestion yes thank you!

**Comment by florentianayuwono** [comment]
For point 2, I tested last time and just now, for t/t and t/x now it only shows contacts with element t and x in it. Perhaps the reason why in yours it shows all contacts for t/x is because there is no contact with x in it?

**Comment by florentianayuwono** [comment]
For point 1, yes I will update the UG, thanks for the suggestion :D

**Comment by florentianayuwono** [comment]
OK now for point 2 I now disable the reduction relevant search if the after reduction length is 0, that way it will prevent the returning all contacts. What do you think?

File: docs/UserGuide.md [hunk 2]
```diff
@@ -105,8 +108,8 @@ Format: `search [CONDITION] [n/NAME] [p/PHONE] [e/EMAIL] [a/ADDRESS] [t/TAG]…`
 * `search and n/NAME p/PHONE…` searches for contacts that match all the given information.
 * `search n/NAME p/PHONE…` and `search and n/NAME p/PHONE…` will return the same contacts that match all the given information.
 * `search or t/TAG1 t/TAG2…` searches for contacts that contain any of the given tags.
-* The search using `n/NAME` is case-insensitive. e.g `hans` will match `Hans`.
-* Displays a list of relevant contacts related to the search query if no search result available.
+* The search using `n/NAME` is case-insensitive. e.g., `hans` will match `Hans`.
```
Comment (Eugene-Ong-W-X) [inline docs/UserGuide.md]: Didn't we agree that `e.g.,` is an abomination? 😆 
```suggestion
* The search using `n/NAME` is case-insensitive. (e.g. `hans` will match `Hans`).
```

**Review by Eugene-Ong-W-X — COMMENTED** [review]
Great work resolving point 2! Tested it and now, 1 character nonexistent tag searches like `search t/x` returns 0 results. 1 small nitpick left on the documentation for you to consider, otherwise everything LGTM 👍

File: docs/UserGuide.md (hunk 2 above)
Comment (florentianayuwono) [reply docs/UserGuide.md]: Oh yaa haha thanks for that ;)

**Pull request merged**